src/core/ext/transport/chttp2/transport/hpack_parse_result.h

```diff
@@ -300,25 +317,31 @@ class HpackParseResult {
     HpackParseStatus status_;
   };
 
-  StatusWrapper status_;
-  union {
-    // Set if status == kInvalidMetadata
-    ValidateMetadataResult validate_metadata_result_;
-    // Set if status == kVarintOutOfRange
-    VarintOutOfRange varint_out_of_range_;
-    // Set if status == kInvalidHpackIndex
-    uint32_t invalid_hpack_index_;
-    // Set if status == kHardMetadataLimitExceeded or
-    // kSoftMetadataLimitExceeded
-    MetadataLimitExceeded metadata_limit_exceeded_;
-    // Set if status == kHardMetadataLimitExceededByKey or
-    // kHardMetadataLimitExceededByValue
-    MetadataLimitExceededByAtom metadata_limit_exceeded_by_atom_;
-    // Set if status == kIllegalTableSizeChange
-    IllegalTableSizeChange illegal_table_size_change_;
+  struct HpackParseResultState {
+    explicit HpackParseResultState(HpackParseStatus incoming_status)
+        : status(incoming_status) {}
+    StatusWrapper status;
+    union {
+      // Set if status == kInvalidMetadata
+      ValidateMetadataResult validate_metadata_result;
+      // Set if status == kVarintOutOfRange
+      VarintOutOfRange varint_out_of_range;
+      // Set if status == kInvalidHpackIndex
+      uint32_t invalid_hpack_index;
+      // Set if status == kHardMetadataLimitExceeded or
+      // kSoftMetadataLimitExceeded
+      MetadataLimitExceeded metadata_limit_exceeded;
+      // Set if status == kHardMetadataLimitExceededByKey or
+      // kHardMetadataLimitExceededByValue
+      MetadataLimitExceededByAtom metadata_limit_exceeded_by_atom;
+      // Set if status == kIllegalTableSizeChange
+      IllegalTableSizeChange illegal_table_size_change;
+    };
+    std::string key;
+    mutable absl::optional<absl::Status> materialized_status;
   };
-  std::string key_;
-  mutable absl::optional<absl::Status> materialized_status_;
+
+  std::shared_ptr<HpackParseResultState> state_ = nullptr;
```

Comment:
If this needs to be a shared pointer, let's prefer `RefCountedPtr` - it saves a decent number of bytes.

Reply:
Done.